### PR TITLE
creating Room endpoint

### PR DESCRIPTION
**Pull request by desiggancheva**

### CODE REVIEW

File: DeskReserve/Controllers/RoomsController.cs
Comment (MartinPetrakiev) [inline DeskReserve/Controllers/RoomsController.cs:1]: Although the logic in each endpoint of the controller is not complex, try separating it into a service.

File: DeskReserve/Data/DBContext/Entity/Room.cs
Comment (MartinPetrakiev) [inline DeskReserve/Data/DBContext/Entity/Room.cs:1]: Try including more Data Annotations to the entity. For example, Default Value, Required...

File: DeskReserve/Controllers/RoomsController.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+
+namespace DeskReserve.Controllers
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class RoomsController : ControllerBase
+    {
+        private readonly ApplicationDbContext _context;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: DbContext doesn't belong here.

File: DeskReserve/Controllers/RoomsController.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        public RoomsController(ApplicationDbContext context)
+        {
+            _context = context;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: What if the context is null ?

File: DeskReserve/Controllers/RoomsController.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        // GET: api/Rooms
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: NO point of this comment.
Remove all from the Rest methods

File: DeskReserve/Controllers/RoomsController.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                return NotFound();
+            }
+
+            _context.Rooms.Remove(room);
+            await _context.SaveChangesAsync();
+
+            return NoContent();
+        }
+
+        private bool RoomExists(Guid id)
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: all other endpoints are asynchronous, this one ?

File: DeskReserve/Controllers/RoomsController.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                return NotFound();
+            }
+
+            return room;
+        }
+
+        // PUT: api/Rooms/5
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPut("{id}")]
+        public async Task<IActionResult> PutRoom(Guid id, Room room)
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Please provide Request objects for Save and update of entity!

File: DeskReserve/Controllers/RoomsController.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            if (id != room.RoomId)
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: ?

File: DeskReserve/Controllers/RoomsController.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            {
+                return BadRequest();
+            }
+
+            _context.Entry(room).State = EntityState.Modified;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Is there a purpose for keeping track of the entity modified state?

File: DeskReserve/Controllers/RoomsController.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                if (!RoomExists(id))
+                {
+                    return NotFound();
+                }
+                else
+                {
+                    throw;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: throw what ?

File: DeskReserve/Controllers/RoomsController.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+        // POST: api/Rooms
+        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
+        [HttpPost]
+        public async Task<ActionResult<Room>> PostRoom(Room room)
+        {
+            _context.Rooms.Add(room);
+            await _context.SaveChangesAsync();
+
+            return CreatedAtAction("GetRoom", new { id = room.RoomId }, room);
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: GetRoom isn't the correct action ?

File: DeskReserve/Domain/RoomService.cs [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+	public class RoomService
+	{
+		private readonly ApplicationDbContext _dbContext;
+
+		public RoomService(ApplicationDbContext dbContext)
+		{
+			_dbContext = dbContext ?? throw new ArgumentNullException(nameof(dbContext));
+		}
+
+		public List<Room> GetAllDogs()
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: IEnumarable

Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: Get all dogs ?!?!?!?!

File: DeskReserve/Domain/RoomService.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+		{
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs:20]: Make it async

File: DeskReserve/Controllers/RoomsController.cs [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+
+namespace DeskReserve.Controllers
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Make sure to hit at least 90% coverage

File: DeskReserve/Domain/RoomService.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+﻿using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+
+namespace DeskReserve.Domain
+{
+	public class RoomService
+	{
+		private readonly ApplicationDbContext _dbContext;
+
+		public RoomService(ApplicationDbContext dbContext)
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: Please register this service to the IOC container

File: DeskReserve/Controllers/RoomsController.cs [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+{
+    [Route("api/[controller]")]
+    [ApiController]
+    public class RoomsController : ControllerBase
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: RoomController use singular

File: DeskReserve/Data/DBContext/Entity/Room.cs [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace DeskReserve.Data.DBContext.Entity
+{
+	[Table("Room")]
+	public class Room
+	{
+		[Key]
+		public Guid RoomId { get; set; }
+		public int RoomNumber { get; set; }
+		public int FloorId { get; set; }
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Data/DBContext/Entity/Room.cs]: Is the FLoor Guid. Also you can fetch the desired relation entity with EntityFramework.

File: DeskReserve/Domain/RoomService.cs [hunk 16]
```diff
@@ -0,0 +1,20 @@
+﻿using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+
+namespace DeskReserve.Domain
+{
+	public class RoomService
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: There is this magical thing called Interface!

File: DeskReserve/Controllers/RoomsController.cs [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        public async Task<ActionResult<Room>> GetRoom(Guid id)
+        {
+            var room = await _context.Rooms.FindAsync(id);
+
+            if (room == null)
+            {
+                return NotFound();
+            }
+
+            return room;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Always specify the return type , in this situation return Ok(room), Be consistent in all controller method

File: DeskReserve/Domain/RoomService.cs (hunk 13 above)
Comment (GeorgiBorisovP3) [reply DeskReserve/Domain/RoomService.cs]: And use it!

**Review by GeorgiBorisovP3 — CHANGES_REQUESTED** [review]

File: DeskReserve/Controllers/RoomsController.cs [hunk 18]
```diff
@@ -0,0 +1,108 @@
+﻿using System;
```
Comment (Dolen-Chalgar) [inline DeskReserve/Controllers/RoomsController.cs]: Where is the creation script?

**Review by Dolen-Chalgar — CHANGES_REQUESTED** [review]

File: DeskReserve/Controllers/RoomsController.cs (hunk 14 above)
Comment (GeorgiBorisovP3) [reply DeskReserve/Controllers/RoomsController.cs]: Still pending !

File: DeskReserve/Controllers/RoomsController.cs [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using DeskReserve.Data.DBContext;
+using DeskReserve.Data.DBContext.Entity;
+using DeskReserve.Domain;
+using System.Net.Mime;
+using System.Drawing;
+using DeskReserve.Exceptions;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Remove redundant usings

File: DeskReserve/Controllers/RoomsController.cs (hunk 3 above)
Comment (GeorgiBorisovP3) [reply DeskReserve/Controllers/RoomsController.cs]: Still pending !

File: DeskReserve/Controllers/RoomsController.cs [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+			{
+				var room = await _roomService.Get(id);
+				result = Ok(room);
+			}
+			catch (EntityNotFoundException ex)
+			{
+				result = NotFound(ex);
+			}
+
+			//var room = await _roomService.Get(id);
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: Remove

File: DeskReserve/Controllers/RoomsController.cs [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+				response = NotFound(ex);
+			}
+
+			return response;
+		}
+
+        /*private bool RoomExists(Guid id)
+        {
+            return _roomService.Rooms.Any(e => e.RoomId == id);
+        }*/
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Controllers/RoomsController.cs]: ?

File: DeskReserve/Data/DBContext/Entity/Room.cs [hunk 22]
```diff
@@ -0,0 +1,17 @@
+﻿using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace DeskReserve.Data.DBContext.Entity
+{
+	[Table("Room")]
+	public class Room
+	{
+		[Key]
+		public Guid RoomId { get; set; }
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Data/DBContext/Entity/Room.cs:10]: setter on id field ?

File: DeskReserve/Data/DBContext/Entity/Room.cs [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+		[Required]
+		public int RoomNumber { get; set; }
+		[Required]
+		public Guid FloorId { get; set; }
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Data/DBContext/Entity/Room.cs:14]: Instead of floorId Get the Entity instead

File: DeskReserve/DeskReserve.csproj [hunk 24]
```diff
@@ -8,17 +8,17 @@
 
   <ItemGroup>
    <!--Swashbucle SWAGGER support for dotnet... -->
-    <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="6.0.16" />
-    <PackageReference Include="Moq" Version="4.20.69" />
-    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
+	  <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="6.0.16" />
+	  <PackageReference Include="Moq" Version="4.20.69" />
+	  <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
```
Comment (GeorgiBorisovP3) [inline DeskReserve/DeskReserve.csproj:13]: Is there any particular reason to change the package version ?

File: DeskReserve/Domain/RoomDto.cs [hunk 25]
```diff
@@ -0,0 +1,10 @@
+namespace DeskReserve.Domain
+{
+	public class RoomDto
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomDto.cs:3]: Split into create and update models

File: DeskReserve/Domain/RoomDto.cs [hunk 26]
```diff
@@ -0,0 +1,10 @@
...
+	{
+		public int RoomNumber { get; set; }
+		//public Guid FloorId { get; set; }
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomDto.cs]: ?

File: DeskReserve/Domain/RoomService.cs [hunk 27]
```diff
@@ -0,0 +1,55 @@
+﻿//using DeskReserve.Data.DBrepository;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: ?

File: DeskReserve/Domain/RoomService.cs [hunk 28]
```diff
@@ -0,0 +1,55 @@
...
+		{
+			var room = await _repository.GetOneAsync(id);
+
+			return room.ToRoomDto();
+
+		}
+
+		public async Task<bool> Update(Guid id, RoomDto roomDto)
+		{
+			var room = await _repository.GetOneAsync(id) ?? throw new EntityNotFoundException();
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Domain/RoomService.cs]: Add meaningful message to the Exception

File: DeskReserve/Mapper/RoomMapper.cs [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+		public static Room ToRoom(this RoomDto RoomDto)
+		{
+			return new Room
+			{
+				RoomNumber = RoomDto.RoomNumber,
+				HasAirConditioner = RoomDto.HasAirConditioner
+			};
+		}
+
+		public static Room UpdateFromDto(this Room Room, RoomDto RoomDto)
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Mapper/RoomMapper.cs:25]: This is not a mapper method!

File: DeskReserve/Program.cs [hunk 30]
```diff
@@ -37,6 +37,9 @@
 
 builder.Services.AddScoped<IFloorRepository, FloorRepository>();
 builder.Services.AddScoped<IFloorService, FloorService>();
+builder.Services.AddScoped<IRoomRepository, RoomRepository>();
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Program.cs:40]: Use Singleton for the Repository

File: DeskReserve/Program.cs [hunk 31]
```diff
@@ -37,6 +37,9 @@
...
+builder.Services.AddScoped<RoomsController, RoomsController>();
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Program.cs]: Remove the controller DI

File: DeskReserve/Repository/IRoomRepository.cs [hunk 32]
```diff
@@ -0,0 +1,18 @@
...
+	{
+		Task<IEnumerable<Room>> GetAllAsync();
+
+		Task<Room> GetOneAsync(Guid id);
+
+		Task<bool> UpdateOneAsync(Room room);
+
+		Task<bool> CreateOneAsync(Room room);
+
+		Task<bool> DeleteOneAsync(Guid id);
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Repository/IRoomRepository.cs]: Use simpler names Get, Update, Create, Delete, why add suffix Async when the return type is Task ?

File: DeskReserve/Repository/RoomRepository.cs [hunk 33]
```diff
@@ -0,0 +1,52 @@
...
+		{
+			await _context.Rooms.AddAsync(room);
+
+			return await _context.SaveChangesAsync() > 0;
+		}
+
+		public async Task<bool> UpdateOneAsync(Room room)
+		{
+			var exict = await _context.Rooms.FindAsync(room.RoomId) ?? throw new EntityNotFoundException();
+			_context.Entry(room).State = EntityState.Modified;
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Repository/RoomRepository.cs:38]: Explain me why we need to track the modify state ?

File: DeskReserve/Repository/RoomRepository.cs [hunk 34]
```diff
@@ -0,0 +1,52 @@
...
+		public async Task<bool> CreateOneAsync(Room room)
+		{
+			await _context.Rooms.AddAsync(room);
+
+			return await _context.SaveChangesAsync() > 0;
+		}
+
+		public async Task<bool> UpdateOneAsync(Room room)
+		{
+			var exict = await _context.Rooms.FindAsync(room.RoomId) ?? throw new EntityNotFoundException();
```
Comment (GeorgiBorisovP3) [inline DeskReserve/Repository/RoomRepository.cs]: exict ?

File: DeskReserve/appsettings.json [hunk 35]
```diff
@@ -5,8 +5,8 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
-  "ConnectionStrings": {
-    "DefaultConnection": "Server=MihailPC;Database=test123;Trusted_Connection=True"
```
Comment (GeorgiBorisovP3) [inline DeskReserve/appsettings.json:9]: Don't commit changes in the appSettings

File: DeskReserve_Tests/RoomControllerTest.cs [hunk 36]
```diff
@@ -0,0 +1,155 @@
+﻿using DeskReserve.Domain;
```
Comment (GeorgiBorisovP3) [inline DeskReserve_Tests/RoomControllerTest.cs:1]: Separate the tests in folder structure similar in the src, Also fix the folder structure in the src

File: DeskReserve_Tests/RoomControllerTest.cs [hunk 37]
```diff
@@ -0,0 +1,155 @@
...
+using Moq;
+using Moq.Language;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (GeorgiBorisovP3) [inline DeskReserve_Tests/RoomControllerTest.cs:8]: Stop commiting redundant usings

File: DeskReserve_Tests/RoomMapper.cs [hunk 38]
```diff
@@ -0,0 +1,79 @@
...
+			Assert.AreEqual(room.RoomNumber, roomDto.RoomNumber);
+			Assert.AreEqual(room.HasAirConditioner, roomDto.HasAirConditioner);
+		}
+
+		/*public static Room UpdateFromDto(this Room Room, RoomDto RoomDto)
+		{
+			Room.RoomNumber = RoomDto.RoomNumber;
+			Room.HasAirConditioner = RoomDto.HasAirConditioner;
+			return Room;
+		}*/
```
Comment (GeorgiBorisovP3) [inline DeskReserve_Tests/RoomMapper.cs:52]: ?

**Review by GeorgiBorisovP3 — APPROVED** [review]

**Review by Dolen-Chalgar — APPROVED** [review]

File: DeskReserve/Controllers/RoomsController.cs (hunk 21 above)
Comment (Dolen-Chalgar) [reply DeskReserve/Controllers/RoomsController.cs]: ?

**Pull request merged**